File: days/day_06/solutions/solution.kt
import java.io.File

data class Problem(val startCol: Int, val endCol: Int, val op: Char)

fun readInputRaw(filePath: String): String = File(filePath).readText()

fun solve(inputData: String): Pair<Long, Long> {
    val lines = inputData.trim().split('\n')
    if (lines.isEmpty()) return Pair(0, 0)
    
    // Find maximum line length and pad all lines
    val maxLen = lines.maxOfOrNull { it.length } ?: 0
    val paddedLines = lines.map { it.padEnd(maxLen) }
    
    // Operator row is the last row
    val opRowIdx = paddedLines.size - 1
    val opRow = paddedLines[opRowIdx]
    val numRows = paddedLines.take(opRowIdx)
    
    // Part 1: Parse horizontally
    var part1Total = 0L
    
    // Find problem boundaries (columns that are all spaces)
    val isSpaceCol = BooleanArray(maxLen) { col ->
        paddedLines.all { col >= it.length || it[col] == ' ' }
    }
    
    // Group columns into problems
    val problems = mutableListOf<Problem>()
    var i = 0
    while (i < maxLen) {
        if (!isSpaceCol[i]) {
            val startCol = i
            while (i < maxLen && !isSpaceCol[i]) i++
            val endCol = i
            // Extract operator for this problem
            val op = opRow.substring(startCol, endCol).firstOrNull { it == '+' || it == '*' }
            op?.let { problems.add(Problem(startCol, endCol, it)) }
        } else {
            i++
        }
    }
    
    // Solve Part 1: Extract numbers horizontally
    for (prob in problems) {
        val numbers = mutableListOf<Long>()
        for (row in numRows) {
            if (prob.endCol > row.length) continue
            val problemStr = row.substring(prob.startCol, prob.endCol).trim()
            val parts = problemStr.split(Regex("\\s+"))
            parts.forEach { part ->
                part.toLongOrNull()?.let { numbers.add(it) }
            }
        }
        
        // Apply operator
        if (numbers.isNotEmpty()) {
            val result = if (prob.op == '+') {
                numbers.sum()
            } else {
                numbers.fold(1L) { acc, n -> acc * n }
            }
            part1Total += result
        }
    }
    
    // Part 2: Parse vertically (columns, right-to-left)
    // Approach: Use Kotlin's functional transformations
    var part2Total = 0L
    
    for (prob in problems) {
        // Extract column strings using functional style
        val colStrings = (prob.startCol until prob.endCol)
            .filter { !isSpaceCol[it] }
            .map { col ->
                numRows.mapNotNull { row -> 
                    row.getOrNull(col)?.takeIf { it.isDigit() || it == ' ' }
                }.joinToString("").trim()
            }
            .filter { it.isNotEmpty() }
            .reversed()  // Right-to-left reading
        
        // Parse numbers
        val numbers = colStrings
            .map { it.filter { c -> c.isDigit() } }
            .filter { it.isNotEmpty() }
            .mapNotNull { it.toLongOrNull() }
        
        // Apply operator
        if (numbers.isNotEmpty()) {
            val result = if (prob.op == '+') {
                numbers.sum()
            } else {
                numbers.fold(1L) { acc, n -> acc * n }
            }
            part2Total += result
        }
    }
    
    return Pair(part1Total, part2Total)
}

fun main() {
    val content = readInputRaw("../data/input.txt")
    val (part1, part2) = solve(content)
    println("Part 1: $part1")
    println("Part 2: $part2")
}

File: days/day_06/solutions/solution.hs
import System.IO
import Data.List
import Data.Char
import Data.Maybe

readInputRaw :: FilePath -> IO String
readInputRaw filePath = readFile filePath

data Problem = Problem { startCol :: Int, endCol :: Int, op :: Char }

solve :: String -> (Integer, Integer)
solve inputData = 
    let inputLines = lines inputData
        trimmedLines = map trim inputLines
    in
    if null trimmedLines
    then (0, 0)
    else
        let maxLen = maximum (map length trimmedLines)
            paddedLines = map (`padRight` maxLen) trimmedLines
            opRowIdx = length paddedLines - 1
            opRow = paddedLines !! opRowIdx
            numRows = take opRowIdx paddedLines
            
            -- Find problem boundaries (columns that are all spaces)
            isSpaceCol = [all (\line -> col >= length line || line !! col == ' ') paddedLines | col <- [0..maxLen-1]]
            
            -- Group columns into problems
            problems = findProblems 0 maxLen isSpaceCol opRow []
            
            -- Part 1: Extract numbers horizontally
            part1Total = sum (map (solveProblem numRows) problems)
            
            -- Part 2: Parse vertically (columns, right-to-left)
            -- Approach: Use pure functional transformations
            part2Total = sum (map (solveProblemPart2 numRows isSpaceCol) problems)
        in
            (part1Total, part2Total)

findProblems :: Int -> Int -> [Bool] -> String -> [Problem] -> [Problem]
findProblems i maxLen isSpaceCol opRow acc
    | i >= maxLen = reverse acc
    | isSpaceCol !! i = findProblems (i+1) maxLen isSpaceCol opRow acc
    | otherwise = 
        let startCol = i
            endCol = findProblemEnd i maxLen isSpaceCol
            op = findOp (drop startCol (take endCol opRow))
        in
            case op of
                Just c -> findProblems endCol maxLen isSpaceCol opRow (Problem startCol endCol c : acc)
                Nothing -> findProblems endCol maxLen isSpaceCol opRow acc

findProblemEnd :: Int -> Int -> [Bool] -> Int
findProblemEnd i maxLen isSpaceCol
    | i >= maxLen = i
    | isSpaceCol !! i = i
    | otherwise = findProblemEnd (i+1) maxLen isSpaceCol

findOp :: String -> Maybe Char
findOp str = find (`elem` "+*") str

solveProblem :: [String] -> Problem -> Integer
solveProblem numRows (Problem startCol endCol op) =
    let numbers = concatMap (parseNumbers startCol endCol) numRows
    in
        if null numbers
        then 0
        else
            let result = case op of
                    '+' -> sum numbers
                    '*' -> product numbers
            in result

parseNumbers :: Int -> Int -> String -> [Integer]
parseNumbers startCol endCol row =
    let problemStr = trim (take (endCol - startCol) (drop startCol row))
        parts = words problemStr
    in
        mapMaybe parseInteger parts

parseInteger :: String -> Maybe Integer
parseInteger s = case reads s of
    [(n, "")] -> Just n
    _ -> Nothing

padRight :: String -> Int -> String
padRight s len = s ++ replicate (len - length s) ' '

solveProblemPart2 :: [String] -> [Bool] -> Problem -> Integer
solveProblemPart2 numRows isSpaceCol (Problem startCol endCol op) =
    let -- Extract column strings (transpose)
        colStrings = reverse [colStr | col <- [startCol..endCol-1],
                                       not (isSpaceCol !! col),
                                       let colStr = trim (map (\row -> row !! col) numRows),
                                       not (null colStr)]
        
        -- Parse numbers from column strings
        numbers = mapMaybe (\s -> parseInteger (filter isDigit s)) colStrings
        
        -- Apply operator
        result = case op of
            '+' -> sum numbers
            '*' -> product numbers
    in if null numbers then 0 else result

trim :: String -> String
trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace

main :: IO ()
main = do
    content <- readInputRaw "../data/input.txt"
    let (part1, part2) = solve content
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2
